File: 2026.1_-_Disciplinas/C01 - PARADIGMAS DE LINGUAGENS DE PROGRAMAÇÃO/Atividade Final - Paradigmas de Linguagens Programação/parte1_imperativo.c
/* =============================================================
 * Parte 1 — Implementacao Imperativa
 * Sistema de Analise de Desempenho Academico
 * Paradigma: Imperativo/Procedural (C)
 * ============================================================= */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ALUNOS 50
#define MAX_NOME 50

/* ----- Estrutura de dados ----- */

typedef struct {
    char nome[MAX_NOME];
    char matricula[20];
    float nota1;
    float nota2;
    float nota3;
    int frequencia;
} Aluno;

/* ----- Funcoes procedurais ----- */

float calcular_media(float n1, float n2, float n3) {
    float media;
    media = (n1 + n2 + n3) / 3.0;
    return media;
}

int verificar_aprovacao(float media, int frequencia) {
    if (media >= 6.0 && frequencia >= 75) {
        return 1; /* aprovado */
    } else {
        return 0; /* reprovado */
    }
}

int ler_arquivo(const char *caminho, Aluno alunos[]) {
    FILE *arquivo;
    char linha[256];
    int total = 0;

    arquivo = fopen(caminho, "r");
    if (arquivo == NULL) {
        printf("Erro ao abrir o arquivo: %s\n", caminho);
        return 0;
    }

    while (fgets(linha, sizeof(linha), arquivo) != NULL) {
        if (strlen(linha) <= 1) {
            continue;
        }
        /* Remove o \n final */
        linha[strcspn(linha, "\r\n")] = '\0';

        /* Parsing manual com sscanf e strtok */
        char *token;
        token = strtok(linha, ";");
        if (token != NULL) strcpy(alunos[total].nome, token);

        token = strtok(NULL, ";");
        if (token != NULL) strcpy(alunos[total].matricula, token);

        token = strtok(NULL, ";");
        if (token != NULL) alunos[total].nota1 = atof(token);

        token = strtok(NULL, ";");
        if (token != NULL) alunos[total].nota2 = atof(token);

        token = strtok(NULL, ";");
        if (token != NULL) alunos[total].nota3 = atof(token);

        token = strtok(NULL, ";");
        if (token != NULL) alunos[total].frequencia = atoi(token);

        total = total + 1;

        if (total >= MAX_ALUNOS) {
            break;
        }
    }

    fclose(arquivo);
    return total;
}

void gerar_relatorio(Aluno alunos[], int total) {
    int i;
    int total_aprovados = 0;
    int total_reprovados = 0;
    float media;
    int aprovado;

    printf("======================================================================\n");
    printf("RELATORIO DE DESEMPENHO ACADEMICO - VERSAO IMPERATIVA (C)\n");
    printf("======================================================================\n");
    printf("%-12s %-12s %-6s %-6s %-6s %-8s %-6s %s\n",
           "Nome", "Matricula", "N1", "N2", "N3", "Media", "Freq", "Situacao");
    printf("----------------------------------------------------------------------\n");

    i = 0;
    while (i < total) {
        media = calcular_media(alunos[i].nota1, alunos[i].nota2, alunos[i].nota3);
        aprovado = verificar_aprovacao(media, alunos[i].frequencia);

        if (aprovado == 1) {
            total_aprovados = total_aprovados + 1;
        } else {
            total_reprovados = total_reprovados + 1;
        }

        printf("%-12s %-12s %-6.1f %-6.1f %-6.1f %-8.2f %-6d %s\n",
               alunos[i].nome,
               alunos[i].matricula,
               alunos[i].nota1,
               alunos[i].nota2,
               alunos[i].nota3,
               media,
               alunos[i].frequencia,
               aprovado ? "Aprovado" : "Reprovado");

        i = i + 1;
    }

    printf("----------------------------------------------------------------------\n");
    printf("Total de alunos: %d\n", total);
    printf("Aprovados: %d\n", total_aprovados);
    printf("Reprovados: %d\n", total_reprovados);
    printf("======================================================================\n");
}

/* ----- Programa principal ----- */

int main(void) {
    Aluno alunos[MAX_ALUNOS];
    int total;

    total = ler_arquivo("alunos.csv", alunos);

    if (total > 0) {
        gerar_relatorio(alunos, total);
    }

    return 0;
}

File: 2026.1_-_Disciplinas/C01 - PARADIGMAS DE LINGUAGENS DE PROGRAMAÇÃO/Atividade Final - Paradigmas de Linguagens Programação/backup/parte1_imperativo.py
# =============================================================
# Parte 1 — Implementação Imperativa
# Sistema de Análise de Desempenho Acadêmico
# Paradigma: Imperativo/Procedural (Python)
# =============================================================

import os

# ----- Funções procedurais -----

def ler_arquivo(caminho):
    """Lê o arquivo CSV e retorna uma lista de listas com os dados."""
    alunos = []
    arquivo = open(caminho, "r", encoding="utf-8")
    linhas = arquivo.readlines()
    arquivo.close()

    i = 0
    while i < len(linhas):
        linha = linhas[i].strip()
        if linha != "":
            partes = linha.split(";")
            nome = partes[0]
            matricula = partes[1]
            nota1 = float(partes[2])
            nota2 = float(partes[3])
            nota3 = float(partes[4])
            frequencia = int(partes[5])
            alunos.append([nome, matricula, nota1, nota2, nota3, frequencia])
        i = i + 1

    return alunos


def calcular_media(nota1, nota2, nota3):
    """Calcula a média aritmética de três notas."""
    media = (nota1 + nota2 + nota3) / 3
    return media


def verificar_aprovacao(media, frequencia):
    """Retorna 'Aprovado' ou 'Reprovado' com base na média e frequência."""
    if media >= 6.0 and frequencia >= 75:
        situacao = "Aprovado"
    else:
        situacao = "Reprovado"
    return situacao


def gerar_relatorio(alunos):
    """Percorre a lista de alunos, calcula médias e imprime o relatório."""
    print("=" * 70)
    print("RELATÓRIO DE DESEMPENHO ACADÊMICO — VERSÃO IMPERATIVA")
    print("=" * 70)
    print(f"{'Nome':<12} {'Matrícula':<12} {'N1':<6} {'N2':<6} {'N3':<6} {'Média':<8} {'Freq':<6} {'Situação'}")
    print("-" * 70)

    total_aprovados = 0
    total_reprovados = 0

    i = 0
    while i < len(alunos):
        nome = alunos[i][0]
        matricula = alunos[i][1]
        nota1 = alunos[i][2]
        nota2 = alunos[i][3]
        nota3 = alunos[i][4]
        frequencia = alunos[i][5]

        media = calcular_media(nota1, nota2, nota3)
        situacao = verificar_aprovacao(media, frequencia)

        if situacao == "Aprovado":
            total_aprovados = total_aprovados + 1
        else:
            total_reprovados = total_reprovados + 1

        print(f"{nome:<12} {matricula:<12} {nota1:<6.1f} {nota2:<6.1f} {nota3:<6.1f} {media:<8.2f} {frequencia:<6} {situacao}")

        i = i + 1

    print("-" * 70)
    print(f"Total de alunos: {len(alunos)}")
    print(f"Aprovados: {total_aprovados}")
    print(f"Reprovados: {total_reprovados}")
    print("=" * 70)


# ----- Programa principal -----

caminho_arquivo = os.path.join(os.path.dirname(os.path.abspath(__file__)), "alunos.csv")
lista_alunos = ler_arquivo(caminho_arquivo)
gerar_relatorio(lista_alunos)
